### ISSUE TITLE
Deploying the same model to multiple databases

### ISSUE THREAD

**Issue by gabrielsmelchor** [otykier/TabularEditor#673]
Good afternoon,

I realize this question may be beyond the scope of this particular forum, but we're facing an interesting challenge. We haven't had much luck finding good resources for this, but I feel confident it can be done.

In short, we have a single database per customer, but the semantic model is the exact same for each of them. We'd like to automate deployment of model updates to each customer using Azure Devops to a dataset hosted in Power BI Premium. I realize one of the steps in the 'You're Deploying it Wrong' series is updating the connection string using a C# script that is prepared in the build pipeline. As we have hundreds of different connection strings, I was wondering if we would need a single repo/pipeline for each deployment, or if we could reuse a single repo/pipeline with a list of connection strings.

Not expecting a detailed response, but if you have any key concepts or resources to share that would be greatly appreciated.

**Comment by otykier** [comment]
This is absolutely possible.

To keep things simple, I would recommend using a PowerShell task in your DevOps pipeline. With PowerShell, you can iterate through a list of connection strings and perform the deployment using Tabular Editor for each of the items in this list, using the technique from the blog to update the connection string inside the model on-the-fly.

I have provided a piece of Powershell code below, that shows how you can perform a single deployment with PowerShell. I leave it up to you to figure out how to place this inside a loop :-)

The script below will:

1. Assign the "SQLDWConnectionString" environment variable from the $ConnectionString powershell variable.
2. Run TabularEditor.exe from the current working directory, loading Model.bim from the same directory.
3. Tabular Editor will execute the script, "SetConnectionStringFromEnv.csx" (same script as used in the blog), to assign the connection string from the environment variable to the data source in the tabular model.
4. Tabular Editor will then perform the deployment $Server and $Database (also PowerShell variables), overwriting existing data sources on this destination.

Of course you may have to adjust the script to point to the correct files in different directories, and also make sure that you assign each of the three variables, $ConnectionString, $Server, $Database for each iteration of the loop. Good luck!

```powershell
# Assign SQLDWConnectionString environment variable (will be picked up by script inside Tabular Editor):
[System.Environment]::SetEnvironmentVariable("SQLDWConnectionString", "$ConnectionString",[System.EnvironmentVariableTarget]::User)

# Run Tabular Editor. Use a script to change connection string before deployment:
$p = Start-Process -FilePath "TabularEditor.exe" -NoNewWindow -PassThru -Wait -ArgumentList "Model.bim -S `"SetConnectionStringFromEnv.csx`" -D `"$Server`" `"$Database`" -O -C -V -E -W"

# Pass Tabular Editor's ExitCode to the caller, in case an error occurred:
exit $p.ExitCode
```

**Comment by gabrielsmelchor** [comment]
@otykier  Thanks a million for the provided snipped and instruction. Super happy to say that even though we were new to Azure Devops, Tabular Editor, and Powershell, we've managed to set up CI/CD solution that allows us to deploy a consistent semantic model to hundreds of clients in a Power BI Premium workspace with a single release. Granted, we're waiting for general release of XMLA write to finish testing, but I feel confident with what we have so far, which is a confirmed model deployment to multiple workspaces with a single release.

To others reading this who may be in the position I was in: the answer to the looping solution was to use Powershell. In short, we use the getGroups REST API call to return an array in Powershell, which we then looped over using foreach. We did some other tasks in this loop, but that's the gist of it. As an aside, we thought we had to learn XMLA (in addition to all of the other new stuff), but turns out we didn't - not yet anyway. Powershell and Tabular Editor is enough to accomplish this dream.

Thanks again.
